src/registry.rs

```diff
@@ -0,0 +1,337 @@
+pub trait ObjectStoreRegistry: Send + Sync + std::fmt::Debug + 'static {
+    /// Register a new store for the provided store URL
+    ///
+    /// If a store with the same URL existed before, it is replaced and returned
+    fn register(&self, url: Url, store: Arc<dyn ObjectStore>) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Resolve an object URL
+    ///
+    /// If [`ObjectStoreRegistry::register`] has been called with a URL with the same
+    /// scheme, and authority as the object URL, and a path that is a prefix of the object
+    /// URL's, it should be returned along with the trailing path. Paths should be matched
+    /// on a path segment basis, and in the event of multiple possibilities the longest
+    /// path match should be returned.
+    ///
+    /// If a store hasn't been registered, an [`ObjectStoreRegistry`] may lazily create
+    /// one if the URL is understood
+    ///
+    /// For example
+    ///
+    /// ```
+    /// # use std::sync::Arc;
+    /// # use url::Url;
+    /// # use object_store::memory::InMemory;
+    /// # use object_store::ObjectStore;
+    /// # use object_store::prefix::PrefixStore;
+    /// # use object_store::registry::{DefaultObjectStoreRegistry, ObjectStoreRegistry};
+    /// #
+    /// let registry = DefaultObjectStoreRegistry::new();
+    ///
+    /// let bucket1 = Arc::new(InMemory::new()) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("s3://bucket1/").unwrap();
+    /// registry.register(base, bucket1.clone());
+    ///
+    /// let url = Url::parse("s3://bucket1/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "path/to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket1));
+    ///
+    /// let bucket2 = Arc::new(InMemory::new()) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("https://s3.region.amazonaws.com/bucket").unwrap();
+    /// registry.register(base, bucket2.clone());
+    ///
+    /// let url = Url::parse("https://s3.region.amazonaws.com/bucket/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "path/to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket2));
+    ///
+    /// let bucket3 = Arc::new(PrefixStore::new(InMemory::new(), "path")) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("https://s3.region.amazonaws.com/bucket/path").unwrap();
+    /// registry.register(base, bucket3.clone());
+    ///
+    /// let url = Url::parse("https://s3.region.amazonaws.com/bucket/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket3));
+    /// ```
+    fn resolve(&self, url: &Url) -> crate::Result<(Arc<dyn ObjectStore>, Path)>;
+}
+
+/// Error type for [`DefaultObjectStoreRegistry`]
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+enum Error {
```

Comment:
Should this be `pub`? Otherwise people can't match on the error type, only on a string message. Maybe that is ok

Reply:
No it is intentionally crate private/opaque, it gets converted into `crate::Error` a few lines down. It purely exists to make the error handling code slightly more ergonomic.

Reply:
```suggestion
///
/// Crate private/opaque type to make the error handling code  more ergonomic. 
/// Always converted into `crate::Error` when reported externally. 
enum Error {
```

Reply:
I don't think this is necessary, this pattern is used throughout this crate

Reply:
😢  -- I think code that can be understood locally as much as possible is preferable (aka not have to research existing patterns in codebase). I realize this is an opinion. In my mind if the pattern wasn't clear to one reader, others will have the same question

Reply:
I've incorporated this in https://github.com/apache/arrow-rs-object-store/pull/375/commits/a3110f50177cd75b2acdbac85ee8a1e967f657f2